pkg/docker/docker.go

```diff
@@ -768,19 +892,21 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) error {
 		os.Exit(2)
 	}
 	return interrupt.New(dumpStack, removeContainer).Run(func() error {
-		// Attach to the container.
+		var attachError error
+		// Attach to the container on go thread (different than with fsouza, since it provided a non-blocking attach which we don't seem to have with k8s/engine-api)
+		// still preserve the flow of attaching before starting to handle various timing issues encountered in the past, as well as allow for --run option
 		glog.V(2).Infof("Attaching to container %q ...", containerName)
-		attachOpts := opts.asDockerAttachToContainerOptions()
-		attachOpts.Container = container.ID
-		if _, err = d.client.AttachToContainerNonBlocking(attachOpts); err != nil {
-			glog.V(0).Infof("error: Unable to attach to container %q with options %+v: %v", containerName, attachOpts, err)
-			return err
-		}
+		go func() {
+			attachOpts := opts.asDockerAttachToContainerOptions()
+			if attachError = d.kubeDockerClient.AttachToContainer(container.ID, attachOpts, opts.asDockerAttachToStreamOptions()); attachError != nil {
+				glog.V(0).Infof("error: Unable to attach to container %q with options %+v: %v", containerName, attachOpts, attachError)
+			}
+		}()
```

Comment:
Yeah, I should have added more comments here.  And I'll preface this by saying that I concede an change may be warranted here.

First, on what occurred with the go-dockerclient:
- remember, the go-dockerclient basically had the same situation since attach was non-blocking 
- go-dockerclient attach returned errors either a) when the container id was empty (should not occur for us since we just created the container we are attaching to) and b) the http post returned an error

So, as your comment below shows, you see that the attachError is defined outside the go-routine.  I started down the path of checking it sooner and blocking until it "changed" somehow, then thought it seemed hacky, so I pulled that and pushed the vanilla check down.

Circling back, where I was headed originally was:
1) set attachError to some dummy error
2) after the go-routine is launched, block until attachError is either changed to nil, or something other than our dummy error
3) then move onto start

That at least would in my mind mirror the go-dockerclient behavior

Too hacky?  Or do you have similar behaving alternative?